Rely on YNAB transactions' cleared state to match them

bibz opened the pull request:
This changeset supports entering manual transactions in YNAB prior to
having the connector match up the bank statement. It consists of two
changes:
    
1- Retrieve bunq transactions as far back as the oldest uncleared
   transaction in YNAB; and,
2- Make sure to match any existing transaction that is still
   uncleared.
    
Note that we do not update existing transactions but rather submit
new transactions with a matching import-id to let YNAB handle the
actual matching. This approach gives a clue to the customer as to when
a manually entered transactions gets matched (YNAB shows the "matched icon"
and requires the match to be approved).
    
Fixes: #34

---

This fix allows budgeters to manually enter transactions in YNAB and still get them matched by the script when they show up in bunq.

Note that transactions can be in three states: "cleared", "uncleared", and "reconciled".
I found it was 1) easier to read, and 2) simpler to shuffle the "is the transaction new" logic in `lib/sync.py#129`.

Commit:
Rely on YNAB transactions' cleared state to match them

This changeset supports entering manual transactions in YNAB prior to
having the connector match up the bank statement. It consists of two
changes:

1- Retrieve bunq transactions as far back as the oldest uncleared
   transaction in YNAB; and,
2- Make sure to match any existing transaction that is still
   uncleared.

Note that we do not update existing transactions but rather submit
new transactions with a matching import-id to let YNAB handle the
actual matching. This approach gives a clue to the customer as to when
a manually entered transactions gets matched (YNAB shows the "matched icon"
and requires the match to be approved).

Fixes: #34

wesselt commented on lib/sync.py:
```diff
@@ -126,9 +126,7 @@ def extend_transactions(self, transactions, payments, syncpair):
-            if transaction:
-                transaction["payment"] = p
-            else:
+            if not transaction or transaction["cleared"] == "uncleared":
```
As I read it, this change will try to create a new transaction at YNAB's end for an uncleared transaction that already exists.

I would expect this to bounce off YNAB's duplicate detection. Updating a transaction in YNAB requires a different mechanism (a PATCH if I remember right.) What works with this change that doesn't work without it?

bibz replied:
You read it right.

As I have experienced it, this effectively works as either a PUT (duplicate) or a PATCH, depending.

I double checked the API following what you said and I realise that there is indeed an actual PATCH that would be the better semantic approach here. I'll review my patch later.

bibz replied:
(but in actuality, this patch works as intended and will allow ynab to register the "matched" part of the bunq transaction.)

bibz replied:
Fixed behaviour with new patch. I now leverage the "dirty" flag instead.

wesselt commented on lib/sync.py:
```diff
@@ -164,7 +165,7 @@ def synchronize_account(self, syncpair):
         # Push start date back to latest YNAB entry
-        if not get_all:
+        if not replay_bunq_tx:
             start_dt = min(start_dt, get_last_transaction_date(transactions))
```
Retrieving bunq transactions is quite slow. The "push back" is there to avoid retrieving too many bunq transactions.  For older accounts like mine, that causes a timeout in Amazon.

What do you think of pushing back a number of days instead of fully?

bibz replied:
That's a fair point. Do you think that transforming the parameter from a flag to a date or number of days would go the right way?

wesselt replied:
Overwriting uncleared transactions seems fine to me.

What's the added value of the `--replay-bunq-tx` flag? What does it do besides setting the `--all` flag?

What do you think of using the date of the last uncleared transaction of the past 30 days as start time for retrieving bunq transactions?

bibz replied:
> What do you think of using the date of the last uncleared transaction of the past 30 days as start time for retrieving bunq transactions?

That seems like a cleaner way to do this, will piggy back on the "35 days back".

bibz commented:
Hi @wesselt , do you still have any interest in moving forward with this patchset or shall we close it as a won't-fix?

bibz commented:
@wesselt I changed approach here to leverage the latest YNAB transaction date as you suggested.

It turns out that "clearing" an existing transaction (and then patching it) does not yield the right outcome. YNAB simply takes it that the TX is now cleared, it does not display the "matched" icon that gives a clue as to what happened to the end-user.

So I am back to submitting a new TX with the right import ID, and that works well. (YNAB does not report duplicates either, so I think it's the right way to do it: `Created 2 and patched 0 transactions.`.)

wesselt commented:
Been testing a bit, looking good. Two requests before I merge it:
- in case there is no YNAB transaction found (or only "Starting Balance"), the script should synchronize all entries instead of the last 35 days
- how can you write `get_last_uncleared_transaction_date` without an exception handler?

bibz commented:
> in case there is no YNAB transaction found (or only "Starting Balance"), the script should synchronize all entries instead of the last 35 days

I am not changing the current behaviour here, if there are no YNAB transactions, then the start date is maintained to what it was. Which would be `2000-01-01`.

> how can you write `get_last_uncleared_transaction_date` without an exception handler?

Sure, I can make `next` return the default value if it found nothing.

wesselt commented:
> I am not changing the current behaviour here

The way I read the code in the old situation it would use "2000-01-01" as starting date when no YNAB transaction is found. In the updated version it would use the last 35 days. (This is without specifying the `--all` flag.)

Just confirmed this in test, if I delete and re-create an account in YNAB the new version doesn't sync it from the start.

bibz commented:
Doh, you are completely right!

bibz commented:
The problem with this approach though (returning 2000-1-1), is that when you have no uncleared TX you will have to get *all bunq transactions*.

wesselt commented:
Hmm, thinking a bit. What about letting the date determination ignore "uncleared" transactions? That way it won't sync more bunq transactions than necessary.

Would that work for your use case?

bibz commented:
I'm afraid not, that is the whole point of this changeset: to account for uncleared transaction so that YNAB can match them to the payments.

What about trying first to look for uncleared transactions, and defaulting to the last not uncleared transaction otherwise?
```
def get_last_transaction_date(transactions):
    """Get the last meaningful transaction date.

    Find the date of the first "uncleared" transaction; or the date
    of the last transaction if none of them is "uncleared"; or fall
    back to 2000-01-01 if there is no actual transactions.
    """
    for transaction in transactions:
        if transaction["payee_name"] == "Starting Balance":
            continue
        if transaction["cleared"] == "uncleared":
            return transaction["date"]
        tx_date = transaction["date"]

    return tx_date or "2000-01-01"
```

(Note that I am also fixing a bug in my own algorithm: we want the oldest transaction that is uncleared, not the most recent.)

wesselt commented:
I assume an uncleared transaction is something that you enter manually in YNAB. Then when the synch happens it gets overwritten. So an uncleared transaction must always be newer than the last one that was synched from bunq.

In which situation can there be an uncleared transaction that is older than the latest cleared transaction?

bibz commented:
An uncleared transaction only gets cleared when there is a matching bunq payment.
Remember that we are talking about *manually entered transactions* here. Whether the payments will follow the same order is unknown, it depends on the whole payment chain - not only on bunq and yourself.

For instance I have a transaction in my budget dating back to a few months when I back-ordered a book. Since the order is still pending, the money is still on my bank account and the transaction is "uncleared" in YNAB. But from the budget perspective, the money is gone since August and is not available anymore.

I feel like this is a very typical use-case of YNAB: you enter transactions as they happen, the bank payments are less important (especially when banks can take a few days to process things).
But of course you might also rely 100% on your bank transactions to update your budget, which would still be handled fine in this case.

wesselt commented:
Interesting!  When I order a book my bunq account is charged right when I order. The situation you describe doesn't occur for me at all.

Thinking about your reply, I think it helps to distinguish between the payment time and the bunq transaction time. The later is the time the transaction shows up in bunq. It's also the timestamp the code uses to upload to YNAB.

Taking your example, say you enter a book payment in YNAB in January. Then the bunq payment comes through in February. The bunq2ynab code will upload a transaction date in February. Won't YNAB see this as two different transactions?

bibz commented:
> Interesting! When I order a book my bunq account is charged right when I order. The situation you describe doesn't occur for me at all.

It depends on the shop you buy the book from, sometimes they charge you when the good is shipped/delivered to the outlet.

> Thinking about your reply, I think it helps to distinguish between the payment time and the bunq transaction time. The later is the time the transaction shows up in bunq. It's also the timestamp the code uses to upload to YNAB.
> 
> Taking your example, say you enter a book payment in YNAB in January. Then the bunq payment comes through in February. The bunq2ynab code will upload a transaction date in February. Won't YNAB see this as two different transactions?

Indeed. If I remember correctly, YNAB will try to match bank transactions give or take *ten days* from when the YNAB transaction is supposed to happen (could also be a future YNAB transaction that actually happened a few days earlier).
I think it is not unreasonable to let YNAB handle the matching part for us, and so we simply submit the transactions based on the facts given by the bank.

javydekoning commented:
What @bibz is saying is very common. I frequently have this as well, I purchase something and pay using Paypal. But Paypal doesn't charge my Bunq account until the item is delivered (hence I add it manually to YNAB as uncleared).

You could do the same for recurring payments, all-though I don't do that, as I just budget for those.

wesselt committed:
More efficient uncleared transaction support

Payments from bunq arrive with increasing timestamps.  So a new bunq
transaction is later than the last cleared payment.  YNAB will take care
of matching an older uncleared transaction.

wesselt commented:
@javydekoning Ah, I set up PayPal as a separate account where I enter transactions manually. Your way sounds easier!

@bibz I tested your latest version. An old uncleared transaction would cause a lot of API traffic. To fix this I modified the code to ignore uncleared transactions. I think this should work, but it's surprisingly complicated.

How does this version work for you?

bibz commented:
You are right, we are getting to a point where it certainly is not trivial to reason about the sync logic.

How would you feel about having some tests for that?
We could for instance 1) write some basic regression tests for the current behaviour, and 2) require additional tests to illustrate a few transaction flows and how they would be handled by new PRs.
(Not sure I can do that in the coming weeks, it is holiday period after all, but I think that would help bring clarity to the use-cases we are discussing and the app in general.)

wesselt commented:
Some tests would be great!  Certainly open to that.

I've been testing commit dfe562d for the past few days and it works fine for me. How does it work for your use case?

wesselt committed:
Auto-sync back-off on error

wesselt committed:
Populate before trying to set up callback

bibz commented:
Thanks for merging @wesselt !
Sorry for the hiatus, I was busy elsewhere. I didn't get to write tests and somehow my problem shows up again.
I will eventually take another look at the logic, but for now I will reconcile things manually.
